### ISSUE TITLE
Not understanding class IndexRanker()

### ISSUE THREAD

**Issue by tdieu29** [stanford-futuredata/ColBERT#66]
Hi @okhat,

I'm having quite a hard time understanding what is being done in class IndexRanker() and I could really use your help. 
It looks like doclens and strides play an important role in determining the output_scores here but I don't understand why we are taking doclens into consideration when calculating the scores of the retrieved pids.

Among some of my questions are:
1. Why are we calculating the strides?

2. Why do we need to compare the doclens of the retrieved pids with the strides?
`assignments = (doclens.unsqueeze(1) > torch.tensor(self.strides).unsqueeze(0) + 1e-6).sum(-1)`

3. Why did we create new views of the embeddings tensor and use some cumulative doclens values as indexes to select values from those new views of the embeddings tensor? 

`def _create_views(self, tensor):`
`        views = []`

`        for stride in self.strides:`
`            outdim = tensor.size(0) - stride + 1`
`            view = torch.as_strided(tensor, (outdim, stride, self.dim), (self.dim, self.dim, 1))`
`            views.append(view)`

`        return views`

`D = torch.index_select(input=views[group_idx], dim=0, index=group_offsets_uniq, out=D_buffers[group_idx][:D_size])`

Thanks a lot!

**Comment by okhat** [comment]
Hey @tdieu29 !

These are all efficiency-oriented snippets of code. The document lengths don't matter for the actual modeling/scoring.

Nonetheless, the doclens are critical for batching the MaxSim score computations. Since each document has its own length, we can't store the index as a 3D tensor. Instead, we store a flattened 2D matrix that packs all documents. The cumulative doclens constitute the offset of each document in the 2D matrix. For instance, doc-0 starts at position zero, whereas doc-1 might start at position 90 if doc-0 has 90 embeddings.

Does this help?

**Comment by tdieu29** [comment]
Hi @okhat,

That totally makes sense! I think I was also able to answer all of my questions. What a clever approach you took!
